llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -2275,9 +2275,18 @@ SDValue DAGCombiner::visitPTRADD(SDNode *N) {
     SDValue Z = N1;
 
     SDValue Add = DAG.getNode(ISD::ADD, DL, IntVT, {Y, Z});
+    // Calling visit() can replace the Add node with ISD::DELETED_NODE if there
+    // aren't any users. Therefore, we created the reassociated result now
+    // regardless of whether we end up returning it since otherwise we might end
+    // up using an invalid node in the checks below.
+    SDValue Reassociated = DAG.getPointerAdd(DL, X, Add);
     SDValue VisitedAdd = visit(Add.getNode());
-    if (VisitedAdd)
-      Add = VisitedAdd;
+    // visit() might RAUW the SDNode (and therefore update the PTRADD argument),
+    // so reload it now.
+    LLVM_DEBUG(dbgs() << "visitPTRADD() reassociated:";
+               Reassociated.dump(&DAG));
+    Add = Reassociated.getOperand(1);
```

Comment:
Should this not be equivalent to the old `if (VisitedAdd) Add = VisitedAdd`?

Reply:
Ideally yes, but unfortuanately there are some cases where visitAdd() does not return the new node. This happens e.g. with the KnowBits fold that is being hit by this test:
```
  if (SimplifyDemandedBits(SDValue(N, 0)))
    return SDValue(N, 0);
```

Reply:
I did consider changing that function to return the new Node, but it seems to be used that way in lots of folds and I don't have time to investigate any resulting test failures.

Reply:
Aha, it seems other places do things like:
```
      SDValue Tmp = visitXOR(N.getNode());
      // No simplification done.
      if (!Tmp.getNode())
        break;
      // Returning N is form in-visit replacement that may invalidated
      // N. Grab value from Handle.
      if (Tmp.getNode() == N.getNode())
        N = XORHandle.getValue();
      else // Node simplified. Try simplifying again.
        N = Tmp;
```
so maybe we want something like:
```
SDValue VisitedAdd = visit(Add.getNode());
if (VisitedAdd) {
  if (VisitedAdd.getNode() == Add.getNode()
    Add = Reassociated.getOperand(1);
  else
    Add = VisitedAdd;
}
```

Reply:
That should also work, but isn't that equivalent to always calling `getOperand(1);`?

Reply:
Many visit functions don't do any in-place modification, they just return a new DAG. DAGCombiner::Run is what does the ReplaceAllUsesWith in the general case.

Reply:
Yes, I realized that shortly after my comment. Now updated to add a test that also tests this case.